sdk/ai/azure-ai-inference/src/main/java/com/azure/ai/inference/ChatCompletionClientTracer.java

```diff
@@ -0,0 +1,606 @@
+            traceChoiceEvent(FINISH_REASON_ERROR, OffsetDateTime.now(ZoneOffset.UTC), span);
+            tracer.end(null, throwable, span);
+            return Mono.empty();
+        };
+
+        final Function<StreamingChatCompletionsState, Mono<Void>> asyncCancel = resource -> {
+            final Context span = resource.span;
+            tracer.setAttribute("error.type", "cancelled", span);
+            traceChoiceEvent(FINISH_REASON_CANCELED, OffsetDateTime.now(ZoneOffset.UTC), span);
+            tracer.end("cancelled", null, span);
+            return Mono.empty();
+        };
+
+        return Flux.usingWhen(resourceSupplier, resourceClosure, asyncComplete, asyncError, asyncCancel);
+    }
+
+    //<editor-fold desc="Private util types, methods">
+    private String rootSpanName(ChatCompletionsOptions completeRequest) {
+        return CoreUtils.isNullOrEmpty(completeRequest.getModel()) ? "chat" : "chat " + completeRequest.getModel();
+    }
+
+    private void traceCompletionRequestAttributes(ChatCompletionsOptions request, Context span) {
+        final String modelId = request.getModel();
+        tracer.setAttribute("gen_ai.operation.name", "chat", span);
+        tracer.setAttribute("gen_ai.system", INFERENCE_GEN_AI_SYSTEM_NAME, span);
+        tracer.setAttribute("gen_ai.request.model", CoreUtils.isNullOrEmpty(modelId) ? "chat" : modelId, span);
+        if (request.getMaxTokens() != null) {
+            tracer.setAttribute("gen_ai.request.max_tokens", request.getMaxTokens(), span);
+        }
+        if (request.getTemperature() != null) {
+            tracer.setAttribute("gen_ai.request.temperature", request.getTemperature(), span);
+        }
+        if (request.getTopP() != null) {
+            tracer.setAttribute("gen_ai.request.top_p", request.getTopP(), span);
+        }
+        if (host != null) {
+            tracer.setAttribute("server.address", host, span);
+            tracer.setAttribute("server.port", port, span);
+        }
+    }
+
+    private void traceCompletionRequestEvents(List<ChatRequestMessage> messages, Context span) {
+        if (!captureContent) {
+            return;
+        }
+        if (messages != null) {
+            for (ChatRequestMessage message : messages) {
+                final ChatRole role = message.getRole();
+                if (role != null) {
+                    final String eventName = "gen_ai." + role.getValue() + ".message";
+                    final String eventContent = toJsonString(message);
+                    if (eventContent != null) {
+                        final Map<String, Object> eventAttributes = new HashMap<>(2);
+                        eventAttributes.put("gen_ai.system", INFERENCE_GEN_AI_SYSTEM_NAME);
+                        eventAttributes.put("gen_ai.event.content", eventContent);
+                        tracer.addEvent(eventName, eventAttributes, OffsetDateTime.now(ZoneOffset.UTC), span);
```

Comment:
you really don't need to provide a timestamp if it's `now` - let the otel do the job for you 

https://github.com/Azure/azure-sdk-for-java/blob/25dfa3de5300583a3986f29f9626efc640434230/sdk/core/azure-core-tracing-opentelemetry/src/main/java/com/azure/core/tracing/opentelemetry/OpenTelemetryTracer.java#L412-L414

Reply:
Nice, all `now` property set have been delegated to OTel.